### PR TITLE
feat(hivemind): completion-driven episodes and semantic routing

### PR DESCRIPTION
**Pull request by sanil-23**
## Summary

Brings TinyHiveMind's completion-driven episodes and semantic routing into
OpenCompany, wired the way `tinyhivemind`'s OpenHuman example wires them:
one semantic router, no second model selector, a deterministic destination
underneath.

Four things land:

- **Completion-driven episodes.** A room can end when every *assigned* member
  has reported `!complete`, instead of only on a quorum it can name. This is the
  one form that is well-defined for a room of *one* — `deliberates()` floors a
  quorum room at two members and `desk_episode` additionally needs the
  configured quorum still reachable, so an asker plus one peer is refused today
  and silently falls back to a single turn.
- **A scheduler seam.** The episode loop asks a `Scheduler` what to do next
  rather than calling `step()` directly. `Quorum` wraps the existing fold;
  `Completion` is the new one. The ~580 lines that build prompts, run turns and
  append rows are untouched.
- **Semantic routing.** `route_desk_message` decides who answers an unaddressed
  desk message; `route` decides who takes a `!broadcast` handoff mid-episode.
  Every decision is written to `routing-trace.jsonl` under the instance root.
- **The responder ladder's model rung** becomes Jev where a routing credential
  is configured, and stays `MeteredSelector` where it is not.

## Why replace the selector rather than layer on it

`MeteredSelector` commits to one id and reports no spread, so this host had to
synthesise `ONE` on the winner and `ZERO` on everyone else. Every rule that reads
a distribution — the `>20%` fan-out above all — is inert against a synthetic one,
so `broadcast` would always route to exactly one teammate however uncertain the
router was. A Jev evaluation carries the real spread; the live trace below is
the first one this host has ever seen.

## Fallback discipline

Every way routing can decline — no credential configured, a transport outage, a
retired model id, a stale roster, a malformed evaluation — falls to the desk's
first other member. An `@mention` short-circuits before any provider call. The
worst case of the new rung is the old rung, which is the rule
`built_in::selector` already states.

The cost of that discipline is that a broken router is **invisible**: the
message still lands, on a sensible teammate, and nothing says a model was never
consulted. That is not hypothetical — see "The pin that did not exist" below.
A fallback is now logged at `warn` with its reason rather than folded into the
success line, and the transport logs the provider's own words at all three
points where a reason still exists.

## Three defects found on the way

- `crate_description` fell through to *"Say one thing to this channel."* for
  `desk_close` after PR #54 renamed the spec to `complete_episode`, so a seat
  would have been told the wrong contract for the call that reports work
  finished. Caught by the belt's own test.
- The two markers were told apart by **audience** — `!broadcast` "when another
  teammate should take work" — which is not a test a seat can apply to its own
  turn, because both markers accept a summary-shaped body. A seat that had
  finished and documented a decision broadcast it anyway (trace below), costing
  a billed routing call, an answer of nobody, and one more turn to write the
  `!complete` it already meant. They now discriminate on what is left **open**,
  which a seat does know.
- `prompt.rs` branched on `turn.phase` to choose a protocol, and `Phase` is
  `Deliberate | Commit` — the two stages of reaching a *quorum* decision. A
  completion turn was therefore handed the deliberation rules ("reply with ONE
  line only… a support with no ^citation does not count") and would have answered
  with a marker line instead of doing the work. Compiled, linted and passed
  every test throughout, because nothing rendered a completion prompt.

## API or behavior changes

- New `typesafe` feature: the System One transport, the only part of the hive
  stack that reaches a network. Off by default.
- `hivemind` ships in the standard build. `JournalSessionLog` was already wired
  in `chat_seed` and only the flag was off; the attributed projection fixes
  agents reading each other's replies as their own voice.
- `vendor/tinyhivemind` bumped `f8e4669` → `cf6db97` (upstream #54 + #55).
  `Utterance::Close` → `CompleteEpisode` (the wire keeps `alias = "close"`, so
  stored rows still read), `Selector` returns a typed `SelectionEvaluation`, and
  `ResponderRequest` gains `minimum_selection_confidence`.
- New `EpisodeEnding::Completed`.
- An operator message to a desk runs completion-driven **unless the desk asked
  to deliberate** — `hive.enabled = true`, which that field's own doc calls
  "what the operator wants". A desk whose manifest reads `quorum = 3,
  blind_round = true` still gets a room that counts; answering it with a
  completion room would be overruling a setting rather than applying a default.
  Group chats keep the quorum path.
- Three tools are withheld from an ordinary member's belt, each on live
  evidence and each a one-line reversal. `SPEECH_TOOLS` narrows from four
  entries to three.
  - `desk_dm`, because it is *attractive* rather than wrong — given a private
    channel a seat takes it and the desk goes dark. Observed: a seat DM'd two
    teammates, got its answer back, and wrote nothing to the desk while the
    operator who asked saw silence. `DmTool` is registered on **no** belt — it
    is withheld everywhere, not relocated to direct conversations — so a
    synchronous peer consult is unavailable company-wide until somebody
    re-registers it, which is a one-line change.
  - `delegate_to_desk` / `delegate_to_teammate`, because on a desk turn no drain
    claims the queue and they refuse inside the model's own turn
    (`reason=drain_unwired`). A refusal costs a whole turn; seen twice in one
    turn on a live run. **This reverses issue #884, D1**, which added
    `delegate_to_teammate` precisely so a desk lead could reach its own
    teammates rather than only other desks. Re-registering either is a one-line
    change once a desk turn claims the drain — that, not the hand-off being
    wrong, is why they are out.
- `!complete` and a bare `!broadcast` no longer leak into the console — a
  completion report renders as a sentence, a handoff marker is stripped from the
  body it carried rather than dropping the message.
- Routing configuration is environment-only for now:
  `OPENCOMPANY_TYPESAFE_API_KEY`, `OPENCOMPANY_TYPESAFE_BASE_URL`,
  `OPENCOMPANY_TYPESAFE_MODEL`.

## Validation

- [x] `cargo fmt --all -- --check`
- [x] `cargo clippy -D warnings` — default, `typesafe`, `openhuman`, and
      `typesafe,openhuman` together
- [x] `cargo test -p opencompany-core --features typesafe --lib` — **5,858
      passed, 0 failed, 1 ignored** on the final commits. The `openhuman` lane
      has not been re-run since the belt narrowed to three entries, which is
      walked by at least one test in `harness::speech_tools`; CI is the first
      signal there.
- [x] **The quorum/routing boundary is pinned.** A `!broadcast` in a
      *deliberating* room must not reach the router — the only place this
      feature can touch the old path. Two mechanisms stop it (the
      `scheduler.completing()` gate and `Quorum::assign` being a trait-default
      no-op) and neither was pinned. The test wires a router that **panics**
      rather than one returning a plan that is then checked: "was not consulted"
      is the property, since a consulted-then-ignored router still costs the
      call the gate exists to save. Verified by deleting the gate and watching
      it fail, then restoring it.
- [x] **The stale-roster guard can now fire.** `roster_version` is the first
      check in `accept_inner` — an evaluation whose version differs from the
      request's is rejected before the domain check or any threshold, so a
      cached answer cannot route to somebody who has left the desk. Both call
      sites passed a *count*, which changes only when the desk changes size, so
      a swap left it identical; and since the provider echoes the number back,
      the check could not fire for any input. Now one shared FNV-1a over the
      candidate ids, called from both, with a test that a swap and a reorder are
      each a new snapshot.
- [x] `scripts/ci/assert-feature-lanes.sh` — the boundary test is gated on
      `typesafe` and the existing lane filters to `hivemind::typesafe`, so it
      would have been #770 exactly: compiled by `Check (--all-features)` and
      executed by nothing. It gets its own lane step, and the row names both
      filters.
- [x] `scripts/ci/assert-rs-source-layout.sh` — inline-test check clean; the
      750-line cap already fails on ~40 pre-existing files, none of them here

`hivemind` now passes five consecutive wide-filter runs. It previously failed
about one in three: `trace` reads the process-wide `OPENCOMPANY_DATA_DIR` and
one test sets it for its own window, so a concurrently-routing test wrote into
the same file and the assertion read somebody else's decision. It passed every
time under a narrow filter, which is how it landed green.

A `Send` bound missing on `Scheduler` only surfaced under
`--features typesafe,openhuman` **together**; each feature alone linted clean.
Feature combinations are their own test surface in this crate.

## What a live run shows

Driven against a real company (`design_studio`, OpenRouter + Composio
configured, a live TypeSafe credential in the host environment).

**Completion-driven episodes work.** A desk answered as a room and terminated
the new way:

```
[hive] a desk answered as a room
  chat=creative_direction  ending=completed  turns=8  failed_turns=0
```

That is the completion scheduler selected, scheduling turns, and ending on
explicit reports rather than on quorum.

**Semantic routing works, and the decision is now reviewable.** From
`routing-trace.jsonl`, on a `!broadcast` handing off an icon stroke-weight
decision for write-up:

```
source     : broadcast
recipients : ["ui_designer"]
plan       : one
model      : jev-1.13.0
confidence : 0.51
disposition: accepted

distribution:
    ui_designer        0.68
    none               0.29
    brand_designer     0.03
```

A real spread, not the synthetic `ONE`/`ZERO` the old rung had to fabricate.
Only `ui_designer` clears the threshold, so the plan is `One` rather than
`Hive` — the `>20%` rule behaving as specified against probabilities a model
actually produced.

The second broadcast of the same episode is the one that found the marker
defect — a finished piece of work handed on with nowhere to go:

```
message    : "icon stroke weight settled at 1.5px, documented at
              agents/ui-designer/icon-stroke-token.md … Ready to merge."
plan       : clarify
recipients : []
    none               0.55
    brand_designer     0.45
```

Jev was right and the seat was wrong. Worth noting that the router declining
cleanly is what made the prompt bug legible at all.

### The same task after the marker contract changed

Same desk, same prompt shape, one token later — "choose the line-height scale
and hand the write-up on":

| | icon stroke (before) | line-height (after) |
|---|---|---|
| turns | 5 | **3** |
| routing calls | 2 | **1** |
| second call's outcome | `clarify`, nobody assigned | — |
| posts from the reporting seat | 3 | **2** |
| protocol text shown to the operator | `!broadcast` at `[194]` | **none** |

The handoff now names the open task rather than the finished one — *"Needs
wiring into the UI kit as CSS custom properties"* — and Jev answered `one` with
0.72 on `ui_designer` at 0.57 confidence. Both seats then closed on the new
test in their own words: *"Nothing left open from my seat."* The receiving seat
spent its turn doing the wiring rather than writing a recap of its own previous
message.

One matched pair, not a body of evidence — see below.

### The pin that did not exist

`DEFAULT_MODEL` was `jev-1.13`, taken from PR #55's prose ("live Jev 1.13
routing") and the TypeSafe adapter's test fixtures. Neither is the service's
list of servable ids, and the endpoint answers that pin with:

```
400 {"detail":{"error_type":"api_usage_error","message":"Unknown model: jev-1.13"}}
```

So *every* broadcast for the whole of that session reached TypeSafe, came back
400, and fell through to the mechanical responder — which picked a sensible
teammate each time. A dead credential and a healthy route were indistinguishable
from outside, and the run read as working routing.

`route_semantic` matches `let Ok(evaluation) = … else` and discards the error,
which is why nothing said so. The transport now logs the provider's body at the
refusal, and the pin is back on `jev-latest` until somebody reads
`model_identity` off a successful call — which the trace above finally makes
possible. It reports **`jev-1.13.0`**: the pin was missing a patch segment, and
`OPENCOMPANY_TYPESAFE_MODEL` takes it for anyone who wants it today.

## What this does not change

**With no routing credential configured, nothing changes at all.**
`router_from_env()` returns `Ok(None)`, the responder ladder keeps
`MeteredSelector`, and `.completing()` is false for every caller that existed
before this branch — so both typesafe surfaces are inert. That covers every
deployment today.

With a credential, be precise about which two things it touches, because only
one of them is gated:

| surface | gated on the completion scheduler? |
|---|---|
| `route_handoff` — who takes a `!broadcast` | **yes**, and pinned by a test |
| `JevSelector` in `choose_responder` — who responds at all | **no** |

So a configured instance gets semantic responder selection **company-wide**,
group chats included — that is the feature, not a leak. What it cannot do is
change how a quorum room then behaves: same `step()`, same fold, same
convergence rules.

`Quorum` wraps `step()` verbatim — same call, same `HiveStep` mapping, and
`commit()` keeps the rule that state advances only after every row is appended.
That is not asserted, it is covered: `deliberation_tests`, `concurrency_tests`
and `moves_grammar_tests` drive real episodes to `Converged` through the new
seam, and `tests/hivemind_e2e.rs` runs the whole path. Those suites predate this
change and pass unmodified — stronger evidence than a unit test of the variant
mapping, because they exercise the ~580 shared lines too: the ones that build
prompts, run turns and append rows, and that a second runner would have had to
duplicate.

The one place the new feature can reach the old path is a `!broadcast` written
in a deliberating room, and that is now pinned by a test using a router that
panics if consulted. Verified by deleting the gate and watching it fail.

## What ships, and what is staged

Two routing entry points were written as a pair, mirroring the reference
runner's `route_message` / `route_broadcast`. **Only the handoff half is
wired.**

- `route` — the mid-episode `!broadcast` handoff. Live.
- `route_desk_message` — the opening route. **Nothing calls it.** Who responds
  to a desk message is answered by `JevSelector` through the responder ladder,
  which also resolves an `@mention` and honours the spend cap; and a completion
  episode opens on the desk's first active member — the lead on a `Lead` desk,
  the shared deterministic fallback on an `Auto` one (issue #1835), which is
  deliberate rather than a placeholder.

It is kept because it is the only thing that can return `RoutingPlan::Hive` — a
primary *and* invitees, opening a room rather than naming one responder, which
`Selector`'s single `choice` cannot express. The case it is staged for is the
**crossing**: a referral resolves `@#desk` to whichever seat is listed first,
and `ReferralConfig::deliberates` says so in its own words — "asking a desk and
being answered by whichever seat happens to be listed first is the thing
deliberation exists to stop." The remedy today is to spend a whole room. A
crossing has no lead contract to respect, which is why it fits there and the
episode opening does not.

Its doc comment says all of this, so three passing tests are not read as
coverage of shipped behaviour.

## Why this is a draft

- **Protocol text still reaches the operator.** Observed live, from a seat
  answering an operator who asked for a design decision:

  > Icon stroke weight decided: 1.5px. Broadcast to the room via `!broadcast` at
  > `[194]` — … The write-up is handed off to whoever picks it up from the
  > broadcast.

  Three leaks in one sentence: the marker name, a transcript row index, and a
  description of the router. Neither is a parsing failure — `readable` rewrites
  only a *leading* marker, exactly as `broadcast_body` parses, so a mid-sentence
  mention is prose and renders verbatim; `[194]` is the sequence number
  `prompt.rs` stamps on every row a seat reads. The cause is that
  `COMPLETE_RULES` names one audience, "read by your teammates", so a seat
  writes to the machine. **Deliberately left for a separate change**: the fix is
  a prompt clause plus a render-side scrub of bare marker tokens, and neither
  belongs in the marker-contract commit.
- **The marker contract has one run behind it.** The before/after above is a
  single matched pair on one desk, not a body of evidence. It moved every number
  the right way — 5 turns to 3, two routing calls to one, three posts from the
  reporting seat to two — and both seats adopted the new framing in their own
  words ("Nothing left open from my seat"). But one run is one run, and a prompt
  change is a request rather than a guarantee.
- **Seats sometimes `@mention` where the protocol asks for `!broadcast`**, which
  routes mechanically and spends no model call. Seen in the pre-fix runs and not
  in the one after, but nothing in this change addresses it and one clean run is
  not evidence it is gone. A prompt problem rather than a routing one, but it
  caps how often the new rung is exercised at all.
- **`EpisodeEnding::Completed` over-reports participants**: `completed_ids`
  returns every participant, including seats seeded complete at open that were
  never assigned. Confirmed again in the run above, whose report named
  `brand_designer` among those who "reported the work done" on a desk where it
  never spoke. Cosmetic in one log line and one chat row.
- **Commit 1 carries the selector swap** as well as the dependency bump and its
  API adaptations; an interactive rebase was not available in this environment.
- **A `desk_dm` consult is invisible to the operator who caused it.** In an
  operator↔agent DM under an ordinary harness setup, the agent may open a
  synchronous consult with its peers — asker seeds, peers deliberate — and the
  operator sees only the final answer, with the exchange that produced it
  nowhere in the channel. That is the same failure that made `desk_dm` worth
  withholding from a desk, one surface along: work happens and the person who
  asked watches silence. It needs to render inline in the DM channel, as a
  thread or a collapsed group, before the tool is trustworthy anywhere. Until
  it does, the one place `desk_dm` is still granted is the place its effect is
  least reviewable.
- **How a routing credential should be provisioned is undecided**, and the
  answer changes what this feature is. Today `router_from_env` reads
  `OPENCOMPANY_TYPESAFE_API_KEY` and nothing else, which makes routing a third
  credential axis independent of the company's inference provider — an instance
  can have a working provider and no router, or the reverse. The three shapes
  worth choosing between:
  - **Managed only**, folded under the platform credential the manager already
    injects, so routing arrives with managed inference and a self-hosted
    instance simply does not have it.
  - **Company-configurable**, provisioned like any other provider, so an
    OpenRouter-backed company can point at its own System One endpoint.
  - **Both**, with the company's own credential outranking the platform's.

  And underneath it, the case this host does not answer at all: **a configured
  provider that cannot serve Jev.** The `None` branch means no credential, and
  it degrades to `MeteredSelector` silently — which, as `jev-1.13` proved
  tonight, is indistinguishable from healthy routing from outside. An
  unsupported provider deserves to be said once at boot, not discovered from a
  distribution that never arrives.

## One borrowed commit

`fix(hivemind): set \`rows\` on the chat path's referral marker` is
cherry-picked from #2369 and is **not** part of this feature. It is here because
`hivemind` shipping in the standard build compiles `runtime/hivemind.rs` on paths
that did not compile it before, and that file does not build on `main`:
`E0063: missing field 'rows'`. Drop it from this branch once #2369 lands.

## Summary by CodeRabbit

- **New Features**
  - Added completion-driven desk episodes that finish when assigned participants explicitly report completion.
  - Added semantic message routing with configurable credential-based routing and deterministic fallback behavior.
  - Added support for broadcast handoffs and clearer completion instructions in episode prompts.
  - Added completion markers and readable transcript rendering for `!complete` and `!broadcast`.

- **Changes**
  - Desk messages now prioritize explicit recipients and support bounded fan-out.
  - Ordinary members retain task spawning but no longer receive desk or teammate delegation tools.
  - Direct-message speech tooling is no longer available in standard speech tool belts.

- **Bug Fixes**
  - Improved episode isolation and completion summaries.
  - Corrected tool description matching for renamed completion actions.

### CODE REVIEW

File: crates/opencompany-core/src/hivemind/completion.rs [hunk 1]
```diff
@@ -0,0 +1,327 @@
...
+}
+
+/// Whether one line carries a marker this module owns.
+///
+/// The cheap pre-check `readable_moves` makes before rewriting anything, so a
+/// transcript with no completion grammar in it is returned untouched.
+#[must_use]
+pub fn is_completion_line(line: &str) -> bool {
+    broadcast_body(line).is_some() || reports_completion(line)
+}
```
[the remark this answers is not in the document]
Comment (sanil-23) [reply crates/opencompany-core/src/hivemind/completion.rs:341]: Fixed in `c896e7d32`, and it led to a second bug from the same root.

`is_completion_line` now uses an exact `is_broadcast` check, so a bare `!broadcast` is recognised and `!broadcasting` is not — as you said, `starts_with` would have been too broad.

Writing the test you suggested surfaced the worse half: `broadcast_body` had the *same* loose `strip_prefix`, so `!broadcasting the results` parsed as a handoff carrying `"ing the results"`. That feeds `reply_broadcast` → `route_handoff`, so a sentence *about* broadcasting would have spent a real routing call on a fragment of its own verb. Both now go through the one boundary check, matching the rule `reports_completion` already applied to `!complete`.

Test added as you asked, plus the `!broadcasting` and mid-sentence cases, and that `readable("!broadcast")` renders a sentence rather than the marker.

File: crates/opencompany-core/src/hivemind/prompt.rs [hunk 2]
```diff
@@ -450,9 +534,14 @@ impl<'a> EpisodePrompt<'a> {
...
-            Phase::Deliberate => self.deliberate_protocol(),
+        let protocol = if self.completing {
+            // Ahead of `phase`, which cannot describe a room that has none.
+            self.complete_protocol()
+        } else {
+            match turn.phase {
+                Phase::Commit => commit_protocol(&self.carried(&standings)),
+                Phase::Deliberate => self.deliberate_protocol(),
+            }
         };
```
[the remark this answers is not in the document]
Comment (sanil-23) [reply crates/opencompany-core/src/hivemind/prompt.rs:545]: Fixed in `c896e7d32`. You were right on both counts — the defect and why the test missed it.

`topic_discipline` and `floor` are now suppressed for a completion turn, the same way the protocol block already was. It is the same root: both branch on `turn.phase`, and `Completion::next` synthesises `Phase::Deliberate` because `Phase` has no value meaning "was assigned". Fixing only the protocol block left the contradiction one paragraph further down, where a room told "nothing here is counted" was then told "an option carries once N different members have backed it".

Your point about the test was the useful part: asserting the *absence* of two strings could never catch text it did not enumerate. It now asserts the grammar itself is absent — `!propose`, `!support`, `!evidence`, `carries once`. I verified it fails when the fix is reverted, so it is a guard rather than a passing assertion.

File: crates/opencompany-core/src/hivemind/typesafe.rs [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+            .filter(|value| !value.is_empty())
+        else {
+            return Ok(None);
+        };
+        let base_url = std::env::var(BASE_URL_ENV)
+            .ok()
+            .map(|value| value.trim().to_owned())
+            .filter(|value| !value.is_empty())
+            .unwrap_or_else(|| DEFAULT_BASE_URL.to_owned());
+        Self::with_base_url(api_key, base_url).map(Some)
```
[the remark this answers is not in the document]
Comment (sanil-23) [reply crates/opencompany-core/src/hivemind/typesafe.rs]: Fixed in `c896e7d32`, with one deliberate difference from the suggestion.

`with_base_url` now refuses an endpoint that is not encrypted, so the key cannot be sent at all — agreed that refusing redirects only protects the second hop and this is the first.

Rather than requiring `https` strictly, it reuses `analytics::config::is_secure_endpoint`, the rule this repo already applies to `OPENCOMPANY_ANALYTICS_ENDPOINT` for the identical reason (a client secret on every request): **`https` anywhere, `http` only to a loopback host**. Two reasons for reusing it rather than writing a second check — a second spelling of "is this safe to send a secret to" is how the two drift, and one of them ends up the lenient one; and it matches on the parsed `Url::host()`, so `http://127.0.0.1.evil.example` is correctly *not* loopback.

That also avoids the test change you flagged: the offline stub tests serve a canned response on an ephemeral `127.0.0.1` port (the shape `chargebee::client_tests` uses) and put nothing on a network, so they keep working without TLS.

Test asserts the refusal names the reason, does not echo the key back, allows loopback, and rejects the look-alike host.

**Comment by sanil-23** [comment]
@coderabbitai review

All three findings from the last pass are addressed in `c896e7d32`, with replies on each thread:

- **`is_completion_line` / bare `!broadcast`** — fixed with an exact marker check. Writing your suggested test surfaced a worse sibling: `broadcast_body` had the same loose `strip_prefix`, so `!broadcasting the results` parsed as a handoff carrying `"ing the results"` and would have spent a real routing call on it. Both now share one boundary check.
- **Quorum blocks on a completion turn** — `topic_discipline` and `floor` are suppressed alongside the protocol block; the test now asserts the grammar itself is absent and fails when the fix is reverted.
- **Cleartext System One endpoint (CWE-319)** — refused at construction, reusing `analytics::config::is_secure_endpoint` (`https` anywhere, `http` only to loopback) rather than a second rule that could drift from it.

Since then the branch merged `main` twice and CI is green on every lane, including `Rust (openhuman, tinymemory)`, which caught two real bugs worth a look:

- a desk answering its **first** message could not open a completion room (`assigned_at: Sequence(0)` is not below a trigger of `Sequence(0)`), and
- an operator message overruled a desk that had explicitly configured `quorum = 3, blind_round = true`.

**Comment by sanil-23** [comment]
@coderabbitai review

File: crates/opencompany-core/src/harness/built_in/brain_tests_part7.rs [hunk 4]
```diff
@@ -141,14 +141,46 @@ async fn two_hive_desk_episodes_in_one_cycle_do_not_fold_into_each_other() {
...
+        .read_from(&company, crate::ports::types::EventSeq::new(0), 512)
+        .await
+        .expect("the journal reads back");
+    let replies: Vec<String> = rows
+        .iter()
+        .filter_map(|row| match &row.event {
+            CompanyEvent::AgentReply { text, .. } => Some(text.clone()),
+            _ => None,
+        })
+        .collect();
```
[the remark this answers is not in the document]
Comment (sanil-23) [reply crates/opencompany-core/src/harness/built_in/brain_tests_part7.rs]: Fixed in `8d2e904fb`. You are right that the global check was not testing what the test claims.

The assertions now carry each reply **with its parent** and check `alpha` under `seq_a`, `beta` under `seq_b`, and — the part the global form could never catch — that neither appears under the other's trigger. Swapped answers would have satisfied the old version completely, which is exactly the folding this test exists to catch.

File: crates/opencompany-core/src/hivemind/completion.rs [hunk 5]
```diff
@@ -0,0 +1,365 @@
...
+    // nothing after it, because it had nothing to complete.
+    //
+    // It also keeps the opening route load-bearing: whoever routing selected is
+    // who the room is waiting on, rather than routing picking a first speaker
+    // and the termination ignoring it.
+    // A room with nobody in it reports Complete on its first pass, which reads
+    // as "the work is done" when what happened is that there was never any. A
+    // desk that cannot seat anybody is a caller bug, and saying so beats
+    // ending silently.
+    if roster.is_empty() {
```
[the remark this answers is not in the document]
Comment (sanil-23) [reply crates/opencompany-core/src/hivemind/completion.rs:163]: Fixed in `8d2e904fb`, taking your diff essentially as written.

`opened` now refuses an empty assignment, next to the empty-roster check and for the same reason: every seat is seeded COMPLETE and the opening assignment reopens those that owe work, so assigning nobody reports `Complete` on the first pass — "the work is done" when none was given out.

Agreed on the distinction that made it worth fixing: `desk_episode` cannot reach it (it falls back to the desk's first member), but `opened` is public and that is its documented contract, so the guard belongs in the function rather than in its one current caller. Test added with a non-empty roster and an empty assignment.

File: crates/opencompany-core/src/hivemind/prompt.rs [hunk 6]
```diff
@@ -463,8 +552,27 @@ impl<'a> EpisodePrompt<'a> {
...
+            if self.completing {
+                String::new()
+            } else {
+                self.topic_discipline(turn.phase)
+            },
+            if self.completing {
+                String::new()
+            } else {
+                self.floor(&standings)
+            },
```
[the remark this answers is not in the document]
Comment (sanil-23) [reply crates/opencompany-core/src/hivemind/prompt.rs:575]: Fixed in `8d2e904fb`. That is the third and last block of the scoreboard, and your reason for why it survives is the right one — `unspoken` stays non-empty when one member takes the floor twice, which is ordinary in a FIFO-scheduled completion room rather than an edge case.

`missing()` is now suppressed for a completion turn alongside `topic_discipline` and `floor`, and the test's forbidden-grammar list gained `!defer` so the instruction cannot come back unnoticed.

**Comment by senamakel** [comment]
we move logic into openhuman+tinyhiveminds in the tinyhivemind repo
